### DISCUSSION TITLE
Aggregating request counts and latency across instances in PromQL?

### DISCUSSION THREAD

**Discussion by marko-asplund** [datadog4s/datadog4s#472] [General]
I've just set up datadog4s in our application for sending http4s metrics to StatsD and I'm now planning to create a Grafana dashboard with plots of successful request counts as well as request latencies, both metrics broken down by path classifier.

AFAIK the query for aggregating request counts across instances would be something like this:

```
sum by(classifier)(increase(myapp_requests_count{status_code="200"}[$__interval]))
```

right?

But how do I correctly aggregate the latency metric that's of type histogram? Something like this?

```
sum by(classifier)(rate(myapp_requests_latency_count{status_code="200"}[$__interval]))
```

**Comment by tomasherman** [comment]
Hi, to be honest, this kind of depends on what you want to measure. When it comes to latencies, i think you want to show something like 'maximal 95th percentile of the latency' or something similar. It doesn't really make much sense summing latencies.

This is what I do for latencies in my dashboard (datadog syntax):
```
max:http_server.requests_latency.95percentile{$env,$operation_id} by {operation_id}
```

Basically, i take the highest of the 95th percentile latency per endpoint (== operationId)

BTW are you saying datadog4s works well with prometheus and grafana? I did not know that :)

**Reply by marko-asplund** [reply]
Hi @tomasherman and thanks for your reply! 👍 

We're using both Datadog as well as Prometheus & Grafana.
Currently we're sending metrics to StatsD and from there it flows to Prometheus and then visualised with Grafana.
I'm able to see the data in Grafana, but wondering about how to properly aggregate and visualize it there.

I'd actually prefer to get the data in Datadog, but haven't managed to do so, and any hints for doing that would be very welcome.
How have you set up the data pipeline for delivering the data in DD?
I tried setting up the DD local host agent and configured the app to send data there, but I didn't see any data.
Does the host agent listen for the metrics by default or is some special config needed?
Is any special config needed to send the data from the agent to DD?
Where should the data be visible in DD?

**Reply by tomasherman** [reply]
See this file: https://github.com/avast/datadog4s/blob/master/datadog-agent.sh ... this is what i use to send data do testing account in ddog ... basically you need to set up datadog agent and then configure datadog4s to send data to that agent 

This should also help: https://docs.datadoghq.com/agent/

**Reply by marko-asplund** [reply]
Thanks! 👍 
My DD agent is able to send other data such as traces to DD and I'm able to see them in DD, but for some reason I don't see metrics sent by datadog4s in DD.
I've our service end set up to send data to DD agent and DD agent set up (account keys etc.) to send data to DD, but just wondering if DD agent needs special config to receive metrics over StatsD protocol as well as to forward that to DD. Or whether the DD end needs some special configuration. Where do you see these metrics in DD? 🤔 
I've been able to verify with `tcpdump` that datadog4s actually does send out the metrics.

**Reply by tomasherman** [reply]
hmm nothing that i know, unfortunately ... you should try datadog support maybe they will be able to help

**Reply by marko-asplund** [reply]
Actually, I'm able to see the metrics in DD now - apparently some learning to do on DD.
Also, need to evaluate whether to use DD or Prometheus/Grafana approach.
Thanks! 🙇

**Reply by tomasherman** [reply]
👍  ... my experience is that if you can afford it, DD is superior :) Anyway, good luck!
